internal/apis/certmanager/types_certificate.go

```diff
@@ -153,8 +153,24 @@ type CertificateSpec struct {
 	// If unset, this defaults to 1/3 of the issued certificate's lifetime.
 	// Minimum accepted value is 5 minutes.
 	// Value must be in units accepted by Go time.ParseDuration https://golang.org/pkg/time/#ParseDuration.
+	// Cannot be set if the `renewBeforePercentage` field is set.
+	// +optional
 	RenewBefore *metav1.Duration
 
+	// `renewBeforePercentage` is like `renewBefore`, except it is a relative percentage
+	// rather than an absolute duration. For example, if a certificate is valid for 60
+	// minutes, and  `renewBeforePercentage=25`, cert-manager will begin to attempt to
+	// renew the certificate 45 minutes after it was issued (i.e. when there are 15
+	// minutes (25%) remaining until the certificate is no longer valid).
+	//
+	// NOTE: The actual lifetime of the issued certificate is used to determine the
+	// renewal time. If an issuer returns a certificate with a different lifetime than
+	// the one requested, cert-manager will use the lifetime of the issued certificate.
+	//
+	// Cannot be set if the `renewBefore` field is set.
+	// +optional
+	RenewBeforePercentage *int
```

Comment:
I think we should discuss whether this should be a string or int value. Let's do that in tomorrow's stand-up.

Reply:
First of all: sorry if you saw a review from me I then deleted!

We've discussed this in standup and the number one concern is that this cannot be an `int`; see [this doc](https://github.com/kubernetes/community/blob/384d42fec0e16929805f42ac9ac983061c2da084/contributors/devel/sig-architecture/api-conventions.md#primitive-types) for justification.

> All public integer fields MUST use the Go int32 or Go int64 types, not int (which is ambiguously sized, depending on target platform). Internal types may use int.

So the last change we need here is to change `int` in these API types to `int32` to match that guide. We've made this mistake already in our resources and we don't want to add any new fields with the plain `int` type.

Could you make that change?

Reply:
> We've discussed this in standup and the number one concern is that this cannot be an int; see [this doc](https://github.com/kubernetes/community/blob/384d42fec0e16929805f42ac9ac983061c2da084/contributors/devel/sig-architecture/api-conventions.md#primitive-types) for justification.

Makes sense - updated to use `int32`!